lib/simulator.py

```diff
@@ -25,23 +26,24 @@ def __init__(self, *args, **kwargs):
         wx.Frame.__init__(self, *args, **kwargs)
 
         self.panel = wx.Panel(self, wx.ID_ANY)
-        self.panel.SetFocus()
+
+        self.panel.SetDoubleBuffered(True)
 
         self.sizer = wx.BoxSizer(wx.VERTICAL)
         self.button_sizer = wx.BoxSizer(wx.HORIZONTAL)
 
         self.button_label = (
-            [_("Speed up"),_('Press + or arrow up to speed up')],
-            [_("Slow down"),_('Press - or arrow down to slow down')],
-            [_("Pause"),_("Press P to pause the animation")],
-            [_("Restart"),_("Press R to restart the animation")],
-            [_("Quit"),_("Press Q to close the simulation window")])
+            [_("Speed up"),_('Press + or arrow up to speed up'), 'animation_speed_up'],
+            [_("Slow down"),_('Press - or arrow down to slow down'), 'animation_slow_down'],
+            [_("Pause"),_("Press P to pause the animation"), 'animation_pause'],
+            [_("Restart"),_("Press R to restart the animation"), 'animation_restart'],
+            [_("Quit"),_("Press Q to close the simulation window"), 'animation_quit'])
         self.buttons = []
         for i in range(0, len(self.button_label)):
             self.buttons.append(wx.Button(self, -1, self.button_label[i][0]))
             self.button_sizer.Add(self.buttons[i], 1, wx.EXPAND)
-            self.buttons[i].Bind(wx.EVT_BUTTON, self.on_key_down)
             self.buttons[i].SetToolTip(self.button_label[i][1])
+            self.buttons[i].Bind(wx.EVT_BUTTON, lambda evt, keycode=self.button_label[i][2]: self.on_key_down(evt, keycode))
```

Comment:
Huh, that's a creative way to get around the fact that you can't do a variable assignment inside a lamba. :)

It's a little confusing to read though.  Can we just do `self.on_key_down(evt, self.button_label[i][2])`?

Actually, I'd prefer if we could rename the function to something like `def self.do_action(self, action)` since it now handles generalized actions and not keypresses at all.  We don't need to pass the `evt` argument in since the function doesn't do anything with the event anymore.